### PR TITLE
[mlir] Add `Operation::dumpPrettyPrinted`

### PR DESCRIPTION
**Pull request by weiweichen**
- [x] Add `Operation::dumpPrettyPrinted` to get more readable print during debugging when the IR may not be able to pass verify yet.

### CODE REVIEW

**Review by River707 — COMMENTED** [review]
I'm not really sure I'm following this API, it seems very narrow and specific. Have you considered instead the possibility of adding an argument to `dump` for providing extra flags?

**Comment by weiweichen** [comment]
> I'm not really sure I'm following this API, it seems very narrow and specific. Have you considered instead the possibility of adding an argument to `dump` for providing extra flags?

Fair enough! Putting in this patch mostly because @lattner mentioned about 👇 to me the other day:
```
I'd love to see an op->dumpPrettyPrinted() method on Operation, which would be suitable for being called by the debugger. If you're interested, I'd very much +1 a patch to upstream MLIR.

It'd set the assumeVerified pretty printing flag, so it would force the "seriously, this is eyebleed, just do the thing I as a human can understand" mode
```

If debugger is ok to run `op->dump(OpPrintingFlags().useLocalScope().assumeVerified())`, "adding an argument to dump for providing extra flags" is probably a better approach?

**Comment by joker-eph** [comment]
> If debugger is ok to run op->dump(OpPrintingFlags().useLocalScope().assumeVerified()), 

I think the idea is that the debugger would run `op->dump(true)`. The prototype for `dump()` would be changed to `void dump(bool assumeVerified = false);`

**Comment by weiweichen** [comment]
> > If debugger is ok to run op->dump(OpPrintingFlags().useLocalScope().assumeVerified()),
> 
> I think the idea is that in the debugger you can type: `op->dump(true)`. The prototype for `dump()` would be changed to `void dump(bool assumeVerified = false);`

Sounds good! Updated the API accordingly.

**Comment by joker-eph** [comment]
LGTM, assuming you checked in the debugger that this work as expected!

File: mlir/include/mlir/IR/Operation.h [hunk 1]
```diff
@@ -320,7 +320,7 @@ class alignas(8) Operation final
 
   void print(raw_ostream &os, const OpPrintingFlags &flags = std::nullopt);
   void print(raw_ostream &os, AsmState &state);
-  void dump();
+  void dump(bool assumeVerified = false);
```
Comment (River707) [inline mlir/include/mlir/IR/Operation.h:323]: If we want to ensure `dump` is available, we should mark it with LLVM_DUMP_METHOD. Helps prevent the symbol from being killed off.

**Review by River707 — APPROVED** [review]

**Comment by lattner** [comment]
For me, the #1 usecase is when working on mlir that is under construction when in LLDB.  E.g. working on building a funciton, it has no terminator, it's like "please just dump it, not in the horrible way".  I'm not sure how well LLDB works with default arguments these days, but this would definitely reduce my suffering on a near daily basis if we had it :-)

File: mlir/include/mlir/IR/Operation.h (hunk 1 above)
Comment (lattner) [reply mlir/include/mlir/IR/Operation.h:323]: +1, please use LLVM_DUMP_METHOD

**Review by lattner — APPROVED** [review]
Awesome, thank you!  +1 with the LLVM_DUMP_METHOD attribute.

**Comment by weiweichen** [comment]
> For me, the #1 usecase is when working on mlir that is under construction when in LLDB. E.g. working on building a funciton, it has no terminator, it's like "please just dump it, not in the horrible way". I'm not sure how well LLDB works with default arguments these days, but this would definitely reduce my suffering on a near daily basis if we had it :-)

lldb doesn't seem to able to pick up the default arguments. I asked our in-house lldb expert @walter-erquinigo who said that "expr eval is based on the information on DWARF, and not on source code, so it doesn’t know your .h files". I don't think I want to change the behavior of `Operation::dump()`. Maybe I shall get back to the `Operation::dumpPrettyPrinted()` API?

**Comment by joker-eph** [comment]
Just a nit on the naming: I would call this `dumpPretty()`  (`Printed` seems redundant, and since you aim at typing this manually in a debugger, less characters won't hurt ;) )

**Comment by River707** [comment]
> Just a nit on the naming: I would call this `dumpPretty()` (`Printed` seems redundant, and since you aim at typing this manually in a debugger, less characters won't hurt ;) )

I'm not sure on the name `pretty`, I don't see how the pretty is configured via the options used, it's closer to a  `dumpVerified`.

**Comment by joker-eph** [comment]
The only effect of "AssumeVerified" is to not disable pretty-printing (that is: not disable custom printers and falling back to generic printer). From a user point of view, it seemed to me `pretty` was somehow OK to describe the visible difference in behavior between `dump()` and `dumpXXXXXX()`, but I'm open to either possibility really.

**Comment by River707** [comment]
> The only effect of "AssumeVerified" is to not disable pretty-printing (that is: not disable custom printers and falling back to generic printer). From a user point of view, it seemed to me `pretty` was somehow OK to describe the visible difference in behavior between `dump()` and `dumpXXXXXX()`, but I'm open to either possibility really.

We have some options (like prettyPrintLocations or something) that I had in my mind that made it weird (even if I'd love to remove that flag)

**Comment by weiweichen** [comment]
Not using default argument (because lldb doesn't understand it) with `Operation::dump()` and adding `Operation::dumpPretty()` since `pretty` feels like similar to other IR printing methods. I don't have a strong opinion on what name to use. Happy to change the API to `dumpVerified` if we like it better (@River707 👀 ).

**Review by lattner — COMMENTED** [review]
great outcome, thank you!

**Pull request merged**